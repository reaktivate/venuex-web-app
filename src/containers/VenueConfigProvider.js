import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { firebaseConnect, isLoaded, isEmpty } from 'react-redux-firebase';
import { FoldingCube } from 'styled-spinkit';
import reactHOC from 'react-hoc';

import { getVenueId, firebaseGet } from '../reducers/core';
import Screen from '../components/Screen';


export const VenueConfig = React.createContext();

export const withVenueConfig = reactHOC(WrappedComponent => props => (
    <VenueConfig.Consumer>
        { venueConfig => <WrappedComponent venueConfig={venueConfig} {...props} /> }
    </VenueConfig.Consumer>
), 'withVenueConfig');


const VenueConfigProvider = ({ venueConfig, venueId, children }) => {
    if (!isLoaded(venueConfig)) {
        return <Screen><FoldingCube /></Screen>;
    }

    if (isEmpty(venueConfig)) {
        return (
            <Screen>
                <h2>Aw, snap! Venue config was not found :(</h2>
                <div>Please double-check your venue ID (&quot;{venueId}&quot;)</div>
            </Screen>
        );
    }

    return (
        <VenueConfig.Provider value={{ id: venueId, ...venueConfig }}>
            {children}
        </VenueConfig.Provider>
    );
};


export default compose(
    firebaseConnect((props, store) => [
        `venues/${getVenueId(store.getState())}`
    ]),
    connect(state => ({
        venueId: getVenueId(state),
        venueConfig: firebaseGet(state, `data/venues/${getVenueId(state)}`)
    }))
)(VenueConfigProvider);


export const __forTesting = { VenueConfigProvider };
