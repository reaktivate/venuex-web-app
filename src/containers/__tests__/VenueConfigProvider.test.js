import React from 'react';
import { shallow, mount } from 'enzyme';

import { __forTesting, withVenueConfig } from '../VenueConfigProvider';

const { VenueConfigProvider } = __forTesting;


const Child = withVenueConfig(({ venueConfig }) => <div>{JSON.stringify(venueConfig)}</div>);
const venueConfig = { name: 'Test Venue', theme: { colors: { primary: '#f0f0f0' } } };

describe('VenueConfigProvider', () => {

    it('config not loaded', () => {
        expect(shallow((
            <VenueConfigProvider venueConfig={undefined} venueId="test_venue">
                <Child />
            </VenueConfigProvider>
        ))).toMatchSnapshot();
    });

    it('empty config', () => {
        expect(shallow((
            <VenueConfigProvider venueConfig={null} venueId="test_venue">
                <Child />
            </VenueConfigProvider>
        ))).toMatchSnapshot();
    });

    it('loaded config', () => {
        expect(shallow((
            <VenueConfigProvider venueConfig={venueConfig} venueId="test_venue">
                <Child />
            </VenueConfigProvider>
        ))).toMatchSnapshot();
    });

    it('withVenueConfig', () => {
        expect(mount((
            <VenueConfigProvider venueConfig={venueConfig} venueId="test_venue">
                <Child />
            </VenueConfigProvider>
        )).find('div')).toMatchSnapshot();
    });
});
