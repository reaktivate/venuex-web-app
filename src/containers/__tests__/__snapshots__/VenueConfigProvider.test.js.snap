// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VenueConfigProvider config not loaded 1`] = `
<styled.div>
  <__styled-spinkit__FoldingCube
    color="#333"
    size={40}
  />
</styled.div>
`;

exports[`VenueConfigProvider empty config 1`] = `
<styled.div>
  <h2>
    Aw, snap! Venue config was not found :(
  </h2>
  <div>
    Please double-check your venue ID ("
    test_venue
    ")
  </div>
</styled.div>
`;

exports[`VenueConfigProvider loaded config 1`] = `
<ContextProvider
  value={
    Object {
      "name": "Test Venue",
      "theme": Object {
        "colors": Object {
          "primary": "#f0f0f0",
        },
      },
    }
  }
>
  <withVenueConfig(Component) />
</ContextProvider>
`;

exports[`VenueConfigProvider withVenueConfig 1`] = `
<div>
  {"name":"Test Venue","theme":{"colors":{"primary":"#f0f0f0"}}}
</div>
`;
